### PR TITLE
SceneClass and SimpleSceneClass

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Was wanting to get these done so that i can eventually do the 3 derived scenes 2d, 3d and interface.

### CODE REVIEW

File: src/w3d/renderer/scene.cpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+#ifdef BUILD_WITH_D3D8
+    for (auto i = 0; i < LightEnvironmentClass::MAX_LIGHTS; ++i) {
+        DX8Wrapper::Set_Light(i, nullptr);
+    }
+#endif
+
+    auto *light_environment = rinfo.m_lightEnvironment;
+    if (light_environment != nullptr) {
```
Comment (tomsons26) [inline src/w3d/renderer/scene.cpp]: should assert rinfo.light_environment==NULL before this and this if case should be  if ( !light_environment  ) as it makes a fallback light if one isn't set

File: src/w3d/renderer/scene.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+            break;
+    }
+}
+
+// 0x00832DF0
+float SimpleSceneClass::Compute_Point_Visibility(RenderInfoClass &rinfo, const Vector3 &point)
+{
+    CastResultStruct res;
+    auto camera_position = rinfo.m_camera.Get_Position();
+    LineSegClass ray{ camera_position, point };
```
Comment (tomsons26) [inline src/w3d/renderer/scene.cpp]: should be LineSegClass ray(camera_position, point);, if the ctor isn't there ask, ill provide the code

File: src/w3d/renderer/scene.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        }
+    }
+    m_releaseList.Reset_List();
+}
+
+// 0x00832B30
+void SimpleSceneClass::Remove_All_Render_Objects()
+{
+    for (auto *robj = m_renderList.Remove_Head(); robj != nullptr; robj = m_renderList.Remove_Head()) {
+        robj->Notify_Removed(this);
```
Comment (tomsons26) [inline src/w3d/renderer/scene.cpp]: original calls SceneClass::Remove_Render_Object()

File: src/w3d/renderer/scene.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/w3d/renderer/scene.cpp]: It is that? do you mean `LineSegClass ray{rinfo.m_camera, point};`

Comment (tomsons26) [reply src/w3d/renderer/scene.cpp]: no, () there's a Vector3, Vector3 ctor

**Comment by OmniBlade** [comment]
@tomsons26 Are there any outstanding changes you were wanting for this PR or does it just need rebasing for merge?

**Comment by tomsons26** [comment]
> @tomsons26 Are there any outstanding changes you were wanting for this PR or does it just need rebasing for merge?

can't remember at this point, only that i don't like that line seg ctor init, its supposed to call the actual ctor not jam the values in.

**Comment by duncanspumpkin** [comment]
It is calling the ctor. It's not an aggregate initialiser if that's what you are getting at. As soon as you define a constructor you can't use aggregate initialisation. Braces and parens do the exact same thing in this context.

File: src/w3d/renderer/scene.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    auto *light_environment = rinfo.m_lightEnvironment;
+    captainslog_dbgassert(light_environment == nullptr, "Light Environment already exists");
+    if (light_environment == nullptr) {
+        static LightEnvironmentClass s_lightEnv;
+        s_lightEnv.Reset({ 0, 0, 0 }, m_ambientLight);
+        for (auto iter = m_lightList.Iterator(); iter; ++iter) {
+            s_lightEnv.Add_Light(*(static_cast<LightClass *>(iter.Get_Obj())));
+        }
+        rinfo.m_camera.Validate_Transform();
+        s_lightEnv.Pre_Render_Update(rinfo.m_camera.Get_Transform_No_Validity_Check());
```
Comment (tomsons26) [inline src/w3d/renderer/scene.cpp]: this would be Get_Transform() since Validate_Transform is called above

**Comment by duncanspumpkin** [comment]
I've rebased and switched back from {} to () and made that other change.

File: src/w3d/renderer/scene.cpp [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+#ifdef BUILD_WITH_D3D8
+    for (auto i = 0; i < LightEnvironmentClass::MAX_LIGHTS; ++i) {
+        DX8Wrapper::Set_Light(i, nullptr);
+    }
+#endif
+
+    auto *light_environment = rinfo.m_lightEnvironment;
+    captainslog_dbgassert(light_environment == nullptr, "Light Environment already exists");
+    if (light_environment == nullptr) {
+        static LightEnvironmentClass s_lightEnv;
```
Comment (OmniBlade) [inline src/w3d/renderer/scene.cpp]: local statics should be prefixed with _ but otherwise named the same as local variables.

**Review by OmniBlade — CHANGES_REQUESTED** [review]
Just need the local static renaming but otherwise I think this is good to merge.

**Review by OmniBlade — APPROVED** [review]

**Pull request merged**